crates/legacy/src/testing/finalization_test.rs

```diff
@@ -0,0 +1,460 @@
+    let (tx_sender, tx_receiver) = broadcast(32);
+
+    let bootstrap_builder_state_id = BuilderStateId::<TestTypes> {
+        parent_commitment: vid_commitment(&[], TEST_NUM_NODES_IN_VID_COMPUTATION),
+        view: ViewNumber::genesis(),
+    };
+
+    let global_state = Arc::new(RwLock::new(GlobalState::new(
+        req_sender,
+        tx_sender.clone(),
+        bootstrap_builder_state_id.parent_commitment,
+        bootstrap_builder_state_id.view,
+        bootstrap_builder_state_id.view,
+        0,
+    )));
+
+    let max_api_duration = Duration::from_millis(100);
+
+    let proxy_global_state = ProxyGlobalState::new(
+        global_state.clone(),
+        BLSPubKey::generated_from_seed_indexed([1; 32], 0),
+        max_api_duration,
+    );
+
+    let (decide_sender, decide_receiver) = broadcast(32);
+    let (da_proposal_sender, da_proposal_receiver) = broadcast(32);
+    let (quorum_proposal_sender, quorum_proposal_receiver) = broadcast(32);
+    let bootstrap_builder_state = BuilderState::<TestTypes>::new(
+        ParentBlockReferences {
+            vid_commitment: vid_commitment(&[], TEST_NUM_NODES_IN_VID_COMPUTATION),
+            view_number: ViewNumber::genesis(),
+            leaf_commit: Commitment::from_raw([0; 32]),
+            builder_commitment: BuilderCommitment::from_bytes([0; 32]),
+        },
+        decide_receiver,
+        da_proposal_receiver,
+        quorum_proposal_receiver,
+        req_receiver,
+        tx_receiver,
+        Default::default(),
+        global_state.clone(),
+        NonZeroUsize::new(TEST_NUM_NODES_IN_VID_COMPUTATION).unwrap(),
+        Duration::from_millis(40),
+        1,
+        Default::default(),
+        Duration::from_secs(1),
+        Default::default(),
+    );
+
+    bootstrap_builder_state.event_loop();
+
+    (
+        proxy_global_state,
+        decide_sender,
+        da_proposal_sender,
+        quorum_proposal_sender,
+        tx_sender,
+    )
+}
+
+/// [process_available_blocks_round] processes available rounds for a given
+/// round. It returns the number of attempts made to get the available blocks
+/// and the result of the available blocks.
+///
+/// By default Consensus will retry 3-4 times to get available blocks from the
+/// Builder.
+async fn process_available_blocks_round(
+    proxy_global_state: &ProxyGlobalState<TestTypes>,
+    builder_state_id: BuilderStateId<TestTypes>,
+    round: u64,
+) -> (
+    usize,
+    Result<Vec<AvailableBlockInfo<TestTypes>>, BuildError>,
+) {
+    let (leader_pub, leader_priv) = BLSPubKey::generated_from_seed_indexed([0; 32], round);
+
+    let current_commit_signature = <BLSPubKey as SignatureKey>::sign(
+        &leader_priv,
+        builder_state_id.parent_commitment.as_ref(),
+    )
+    .unwrap();
+
+    // Simulate Consensus retries
+
+    let mut attempt = 0;
+    loop {
+        attempt += 1;
+
+        let available_blocks_result = proxy_global_state
+            .available_blocks(
```

Comment:
If we want to test the number of attempts as updated in `build_block`, can we call `build_block` directly here? I'm confused about how calling `available_blocks` simulates the retries we want to test.

Maybe this is totally the right implementation. Just wanted to make sure since I'm not certain about this. :)

Reply:
This test is setup to simulate the way consensus interacts with the `Builder` by calling `available_blocks`.  This is also where Consensus will trigger 4 attempts when an error is returned.

Reply:
Makes sense!